crates/polars-ops/src/frame/join/iejoin/mod.rs

```
@@ -377,90 +373,129 @@ unsafe fn materialize_join(
+                .arg_sort(l1_sort_options)
+                .slice(x.null_count() as i64, x.len() - x.null_count());
+
+            let y_ordered_by_x = unsafe { y.take_unchecked(&l1_order) };
+            let l2_sort_options = SortOptions::default()
+                .with_maintain_order(true)
+                .with_nulls_last(false)
+                .with_order_descending(l2_descending);
+            // Get the indexes into l1, ordered by y values.
+            // l2_order is the same as "p" from Khayyat et al.
+            let l2_order = y_ordered_by_x.arg_sort(l2_sort_options).slice(
+                y_ordered_by_x.null_count() as i64,
+                y_ordered_by_x.len() - y_ordered_by_x.null_count(),
+            );
+            let l2_order = l2_order.rechunk();
+            let l2_order = l2_order.downcast_as_array().values().as_slice();
+
+            let row_idx = with_match_physical_numeric_polars_type!(x.dtype(), |$T| {
+                 ie_join_impl_t::<$T>(
+                    slice,
+                    l1_order,
+                    l2_order,
+                    op1,
+                    op2,
+                    x,
+                    y_ordered_by_x,
+                    left_height
+                )
+            })?;
 
-    let mut y = selected_left[1].to_physical_repr().into_owned();
-    y.extend(&selected_right[1].to_physical_repr())?;
-    // Rechunk because we will gather.
-    let y = y.rechunk();
+            Ok(row_idx)
+        }
+    );
 
-    let l1_sort_options = SortOptions::default()
-        .with_maintain_order(true)
-        .with_nulls_last(false)
-        .with_order_descending(l1_descending);
-    // Get ordering of x, skipping any null entries as these cannot be matches
-    let l1_order = x
-        .arg_sort(l1_sort_options)
-        .slice(x.null_count() as i64, x.len() - x.null_count());
+    let result = partial_results
+        .map(|pr: Result<Vec<(IdxSize, IdxSize)>, PolarsError>| pr.unwrap())
```

Uwrap here is not ideal, this will panic rather than return an error that can be handled. Maybe explicitly iterating over the results would be cleaner than try to make this work with `reduce`. That should also solve the problem of the unnecessary extra sort mentioned in the TODO comment

Eg. something like:
```rust
    let mut partial_result_1 = partial_results.next().unwrap()?;
    partial_result_1.sort_unstable();
    while let Some(partial_result_2) = partial_results.next() {
        let mut partial_result_2 = partial_result_2?;
        partial_result_2.sort_unstable();
        let mut partial_result: Vec<(IdxSize, IdxSize)> = Vec::new();
        // ...
        partial_result_1 = partial_result;
    }
```

(need to make `partial_results` above mut, and maybe renaming things after might make this clearer)